liquibase-core/src/main/java/liquibase/command/CommandFactory.java

```diff
@@ -63,13 +62,64 @@ public CommandDefinition getCommandDefinition(String... commandName) throws Ille
+        final Set<CommandStep> pipeline = new LinkedHashSet<>();
+        DependencyUtil.DependencyGraph<CommandStep> pipelineGraph = new DependencyUtil.DependencyGraph<>(
+                p -> { if (p != null) pipeline.add(p); }
+        );
+
+        Collection<CommandStep> allCommandStepInstances = findAllInstances();
+        findDependenciesForCommand(pipelineGraph, allCommandStepInstances, this.filterCommandDefinition(allCommandStepInstances, commandName));
+        pipelineGraph.computeDependencies();
+
+        if (pipeline.isEmpty()) {
+            throw new IllegalArgumentException("Unknown command '" + StringUtil.join(commandName, " ") + "'");
+        } else {
+            pipeline.forEach(p -> {
+                try {
+                    commandDefinition.add(p.getClass().getConstructor().newInstance());
+                } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                    throw new IllegalArgumentException(e);
+                }
+            });
+        }
+    }
+
+    public CommandStep filterCommandDefinition(Collection<CommandStep> allCommandStepInstances, String... commandName) {
+        String joinedCommandName = StringUtil.join(commandName, " ");
+
+        for (CommandStep step : allCommandStepInstances) {
+            final String[][] definedCommandNames = step.defineCommandNames();
+            if (definedCommandNames != null) {
+                for (String[] thisCommandName : definedCommandNames) {
+                    if ((thisCommandName != null) && StringUtil.join(thisCommandName, " ")
+                            .equalsIgnoreCase(joinedCommandName)) {
+                        return step;
+                    }
+                }
+            }
         }
+        throw new IllegalArgumentException("Unknown command '" + joinedCommandName + "'");
+    }
 
+    private void findDependenciesForCommand(DependencyUtil.DependencyGraph<CommandStep> pipelineGraph, Collection<CommandStep> allCommandStepInstances,
+                                            CommandStep step) {
+        if (step.requiredDependencies().isEmpty()) {
+            pipelineGraph.add(null, step);
+        } else {
+            for (Class<?> d : step.requiredDependencies()) {
+                CommandStep provider = whoProvidesClass(d, allCommandStepInstances);
+                pipelineGraph.add(provider, step);
+                findDependenciesForCommand(pipelineGraph, allCommandStepInstances, provider);
+            }
+        }
+    }
 
-        return commandDefinition;
+    private CommandStep whoProvidesClass(Class<?> d, Collection<CommandStep> allCommandStepInstances) {
+        return allCommandStepInstances.stream().filter(cs -> cs.providedDependencies().contains(d))
+                .findFirst().orElseThrow(() -> new RuntimeException("Unable to find CommandStep provider for class " +  d.getName()));
```

Comment:
Please add javadocs and inline documentation to these methods, which appear to be cognitively complex.

Reply:
👍 Some comments here would be very helpful.

Reply:
Done!